### PR TITLE
Handle an etcd watch failure more gracefully.

### PR DESCRIPTION
**Pull request by brendandburns**
Partial fix for #3216.

Need to fix up the tests, but sending early for comments from @lavalamp before he turns into a pumpkin ;)

### CODE REVIEW

**Comment by lavalamp** [comment]
This is the thing @erictune and I were chatting about this morning, I think.

If I understand it, this PR changes watch by auto-watching at the global etcd index when the window has expired. That is an salvageable fix, I'm pretty sure: The client would silently miss changes between the window expiration and the current global index.

Instead, our list needs to return the global etcd index which etcd returns along with our call to list instead of the modified index.

File: pkg/tools/etcd_tools_watch.go [hunk 1]
```diff
@@ -150,21 +151,31 @@ func (w *etcdWatcher) etcdWatch(client EtcdGetSet, key string, resourceVersion u
...
-	_, err := client.Watch(key, resourceVersion, w.list, w.etcdIncoming, w.etcdStop)
+	res, err := client.RawWatch(key, resourceVersion, w.list, w.etcdIncoming, w.etcdStop)
 	if err != nil && err != etcd.ErrWatchStoppedByUser {
+		if IsEtcdEventIndexCleared(err) {
+			resourceIndex := res.Header.Get("X-Etcd-Index")
+			index, parseErr := strconv.ParseUint(resourceIndex, 10, 64)
+			if parseErr != nil {
+				err = fmt.Errorf("Poorly formated etcd-index: %s %v", resourceIndex, parseErr)
+			} else {
+				w.etcdWatch(client, key, index+1)
```
Comment (lavalamp) [inline pkg/tools/etcd_tools_watch.go]: All changes between end of window and current index will be missed!

**Comment by smarterclayton** [comment]
Yeah, the only way to know what the next index to watch at is to have etcd return a no-op event during watch when certain thresholds are passed that contains the new resource version, then pass that through watch.Interface back down to clients, then modify clients to expect the no-op / resource version reset event.

The only other valid solution is that our "watch" loops have to reset to get initial state whenever this class of error is returned.

**Comment by lavalamp** [comment]
> Could we do an etcd get unconditionally, and then filter by ResourceVersion of each resource?

Depending on what you mean by that, yes, but it's vastly inferior to having LIST (maybe GET, too) return the global index instead of the modified index. More complicated and bandwidth intensive.

**Comment by brendandburns** [comment]
I looked, and LIST actually does appear to return the global etcd index:

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/tools/etcd_tools.go#L163
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/tools/etcd_tools.go#L169

I think the actual problem might be in the go-etcd client code itself:
https://github.com/coreos/go-etcd/blob/master/etcd/watch.go#L51

@bcwaldon 
Don't you think that should be:
```go
waitIndex = resp.EtcdIndex + 1
```
?

**Comment by lavalamp** [comment]
> Yeah, the only way to know what the next index to watch at is to have etcd return a no-op event during watch when certain thresholds are passed that contains the new resource version, then pass that through watch.Interface back down to clients, then modify clients to expect the no-op / resource version reset event.

This is a possible solution, but it requires a bunch of new machinery.

Do you get 401 only when you start a watch with an expired window, or also if the window expires on an existing watch?

e.g., if the window is 1000, you try to watch at 100, but the global index is at 1101, you get a 401 error. Do you also get a 401 error if you watch while the global index is at 1099, and then two unrelated changes happen?

If the answer is "no", then I think the solution of returning the global index from list is much better. If the answer is "yes", then I'm ambivilent-- @smarterclayton's solution is better but also a lot more work.

**Comment by lavalamp** [comment]
@brendandburns but modified index should == global index if you just got a new change, hm. I am no longer so sure about what is going on.

**Comment by smarterclayton** [comment]
It's only when you try to watch something that hasn't changed within the window.  

Watching from the global index on list makes the problem occur less frequently, but doesn't remove the problem (it's strictly more correct of course).  If you watch from the global index you're ensuring you start at the front of the window, but if the rate of change on objects you are watching is much lower than the rate of change on the server, you will be listing much more frequently than you truly need to.

I.e. If > 1000 changes to etcd occur in between pod creations, then the scheduler will have to reload all pods every time it wants to schedule.  Watching from the front of the window simply increases the portion of the window you see.

The problem is essentially the two generals problem - etcd knows that no modifications have occurred while you are watching, but the client does not know (not receiving a message != no message sent).  The client needs a positive ACK that the window is changing from the server to avoid having to reconnect (refresh state).

**Comment by lavalamp** [comment]
I still am confused about why etcd can't fix this. Etcd has a list of watches, and knows they haven't been triggered by anything up to the current global index; why can't etcd increment the index for each watch every time the global index increments without triggering the watch?

**Comment by xiang90** [comment]
@lavalamp Can you please send an issue or feature request to etcd repo? Thanks. No one has asked etcd to "fix" a problem. :(

**Comment by lavalamp** [comment]
@xiang90: Happy to do so!

**Comment by erictune** [comment]
closing because it looks like #3240 will be used again.

**Pull request closed without merging**